proposals/csharp-7.0/tuples.md

```diff
@@ -0,0 +1,437 @@
+# Tuples 
+
+This proposal specifies the changes required to the [C# 6.0 (draft) Language specification](../../spec/introduction.md) to support *Tuples* as a new [value type](../../spec/types.md#value-types).
+
+## Changes to [Lexical structure](../../spec/lexical-structure.md)
+
+> The grammar for [Literals](../../spec/lexical-structure.md#literals) is extended to include `tuple_literal`:
+
+```antlr
+literal
+    : boolean_literal
+    | integer_literal
+    | real_literal
+    | character_literal
+    | string_literal
+    | null_literal
+    | tuple_literal // new
+    ;
+
+tuple_literal
+    : '(' tuple_literal_element_list ')'
+    ;
+
+tuple_literal_element_list
+    : tuple_literal_element ',' tuple_literal_element
+    | tuple_literal_element_list ',' tuple_literal_element
+    ;
+
+tuple_literal_element
+    : ( identifier ':' )? expression
+    ;
+```
+
+Note that because it is not a constant expression, a tuple literal cannot be used as default value for an optional parameter.
+
+## Changes to [Types](../../spec/types.md)
+
+> The first paragraph of [Value types](../../spec/types.md#value-types) is replaced with the following text:
+
+A value type is either a `struct` type, an enumeration type, or a tuple type. C# provides a set of predefined struct types called the ***simple types***. The simple types are identified through reserved words.
+
+> The `value_type` grammar is updated to include `tuple_type`:
+
+```antlr
+value_type
+    : struct_type
+    | enum_type
+    | tuple_type // new
+    ;
+
+tuple_type
+    : '(' tuple_type_element_list ')'
+    ;
+    
+tuple_type_element_list
+    : tuple_type_element ',' tuple_type_element
+    | tuple_type_element_list ',' tuple_type_element
+    ;
+    
+tuple_type_element
+    : type identifier?
+    ;
+```
+
+## Additions to [Types](../../spec/types.md)
+
+> The following sections should be added after [Nullable types](../../spec/types.md#nullable-types) (at the end of the current *Value types* section.)
+
+### Tuple types
+
+Tuple types are declared with the following syntax:
+
+```csharp
+public (int sum, int count) Tally(IEnumerable<int> values) { ... }
+
+var t = Tally(myValues);
+Console.WriteLine($"Sum: {t.sum}, count: {t.count}");
+ ```
+
+The syntax `(int sum, int count)` indicates an anonymous data structure with public fields of the given names and types, referred to as *tuple*.
+
+Tuple values can be created using *tuple literals*:
+
+```csharp
+var t1 = (sum: 0, count: 1);
+var t2 = (0, 1);     // field names are optional
+```
+
+Tuples cannot be created with the `new` operator. The `new` operator can be used to initialize arrays of tuples, or nullable tuples:
```

Comment:
>Tuples cannot be created with the `new` operator. [](start = 0, length = 49)

Is this a language rule or a comment about what constructors exist?

Reply:
logged as #3152